src/arangodb.c

```diff
@@ -0,0 +1,1201 @@
+    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 7L);
+    curl_easy_setopt(curl, CURLOPT_NOSIGNAL, 1L);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, cluster_write_callback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, curlinfo);
+    curl_easy_setopt(curl, CURLOPT_USERAGENT, COLLECTD_USERAGENT);
+    curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, curlinfo->curl_errbuf);
+    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
+    curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 3L);
+  
+  } else {
+    WARNING("arangodb plugin: curl_easy_init() failed");
+  } // else
+
+} // start_curl_session
+
+
+//
+// send http request via curl and process the response's JSON content
+//
+static int curl_perform(curlinfo_t * curlinfo, endpoint_t * endpoint, char * page, yajl_val * response) {
+  int ret_val, status;
+  long rc;
+  char * url;
+
+  ret_val = 0;
+  
+  sfree(curlinfo->temp_buffer);
+  curlinfo->buf_size = 0;
+
+  url = ssnprintf_alloc("%s%s:%s%s",
+                        endpoint->url_scheme,
+                        endpoint->host, endpoint->port,
+                        page);
+
+  curl_easy_setopt(curlinfo->curl, CURLOPT_URL, url);
+  status = curl_easy_perform(curlinfo->curl);
+  sfree(url);
+  
+  if (status != CURLE_OK) {
+    ERROR("arangodb plugin: curl_easy_perform failed with status %i: %s (%s)",
+          status, curlinfo->curl_errbuf, url);
+    ret_val = -1;
+  } // if
+  
+  curl_easy_getinfo(curlinfo->curl, CURLINFO_RESPONSE_CODE, &rc);
+
+  // The response code is zero if a non-HTTP transport was used.
+  if ((rc != 0) && (rc != 200)) {
+    ERROR("arangodb plugin: curl_easy_perform failed with "
+          "response code %ld (%s)",
+          rc, url);
+    ret_val = -1;
+  } // if
+
+  *response = NULL;
+  
+  // if curl likes it, see if response appears valid
+  if (0 == ret_val) {
+    *response = yajl_tree_parse(curlinfo->temp_buffer, curlinfo->curl_errbuf, sizeof(curlinfo->curl_errbuf));
+
+    // read result codes within JSON response
+    if (NULL != *response && YAJL_IS_OBJECT(*response)) {
+      bool good;
+      yajl_val code_val, error_val;
+      const char * code_path[] = { "code", (const char *) 0 };
+      const char * error_path[] = { "error", (const char *) 0 };
+      
+      code_val = yajl_tree_get(*response, code_path, yajl_t_number);
+      error_val = yajl_tree_get(*response, error_path, yajl_t_any);
+
+      // code and error need to both be missing, or both be positive
+      
+      good = (NULL == code_val && NULL == error_val)
+        || (YAJL_IS_INTEGER(code_val) && 200 == YAJL_GET_INTEGER(code_val)
+            && YAJL_IS_FALSE(error_val));
+
+      if (!good) {
+        ret_val = -1;
+        WARNING("arangodb plugin: response object indicates error.");
+      } // if
+    } else {
+      ret_val = -1;
+      WARNING("arangodb plugin: response JSON did not parse.");
+    } // else
+  } // if
+
+  return ret_val;
+
+} // curl_perform
+
+
+static size_t cluster_write_callback(void *buf,
```

Comment:
Can you please rename this function into `cluster_curl_callback` to don't confuse.
You don't write nothing into cluster.